### ISSUE TITLE
.NET Core Process APIs not Consistent (or always useful) in all situations

### ISSUE THREAD

**Issue by richlander** [dotnet/corefx#36086]
# .NET Core Process APIs not Consistent (or always useful) in all situations

The Process APIs should return accurate and useful values in all situations. I wrote a [simple app that uses all the memory APIs](https://gist.github.com/richlander/6a88c41e06dc92751fa28e43124796cf) I could find. I am happy to add more if people direct me to them.

Major take-aways:

* Multiple properties return a value of 0 on Unix.
* `Process.MaxWorkingSet` != `memory.limit_in_bytes` when memory limited
* `Process.WorkingSet64` != `memory.usage_in_bytes`

Related issues:

* [PrivateMemorySize64](https://github.com/dotnet/corefx/issues/23449)
* [Process.MaxWorkingSet should read from /sys/fs/cgroup/memory when available](https://github.com/dotnet/corefx/issues/35638)

## On Windows

```console
C:\testapps\memory>dotnet run

***Configuration**
Build: .NET Core 3.0.0-preview4-27514-07
OS: Microsoft Windows 10.0.17763
OS Arch: X64

***Working set****
Process.MinWorkingSet: 204800
Process.MaxWorkingSet: 1413120
Process.PeakWorkingSet64: 15667200
Process.WorkingSet64: 15663104

***Memory*********
Process.PeakPagedMemorySize64: 267993088
Process.PrivateMemorySize64: 267993088
Process.NonpagedSystemMemorySize64: 42288
Process.PagedMemorySize64: 267993088
Process.PagedSystemMemorySize64: 138040

***Virtual memory
Process.VirtualMemorySize64: 2307580309504
Process.PeakVirtualMemorySize64: 2307580309504
```

## On Linux

```console
rich@tofino:~/memory$ dotnet run

***Configuration**
Build: .NET Core 3.0.0-preview4-27514-07
OS: Linux 3.10.105-bsp-1.2 #2 SMP PREEMPT Sat Oct 27 19:24:05 IST 2018
OS Arch: Arm64

***Working set****
Process.MinWorkingSet: 0
Process.MaxWorkingSet: 9223372036854775807
Process.PeakWorkingSet64: 0
Process.WorkingSet64: 18436096

***Memory*********
Process.PeakPagedMemorySize64: 0
Process.PrivateMemorySize64: 0
Process.NonpagedSystemMemorySize64: 0
Process.PagedMemorySize64: 0
Process.PagedSystemMemorySize64: 0

***Virtual memory
Process.VirtualMemorySize64: 2735525888
Process.PeakVirtualMemorySize64: 0

***CGroup*********
CGroup usage: 1644941312
CGroup memory limit: 9223372036854775807
```

## On macOS

```console
Richards-MacBook-Pro:memory rich$ dotnet run

***Configuration**
Build: .NET Core 3.0.0-preview4-27514-07
OS: Darwin 18.2.0 Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64
OS Arch: X64

***Working set****
Process.MinWorkingSet: 0
Process.MaxWorkingSet: 9223372036854775807
Process.PeakWorkingSet64: 0
Process.WorkingSet64: 18923520

***Memory*********
Process.PeakPagedMemorySize64: 0
Process.PrivateMemorySize64: 0
Process.NonpagedSystemMemorySize64: 0
Process.PagedMemorySize64: 0
Process.PagedSystemMemorySize64: 0

***Virtual memory
Process.VirtualMemorySize64: 6893395968
Process.PeakVirtualMemorySize64: 0
```

## On WSL

```console
rich@phantomlake:/mnt/c/testapps/memory$ ~/dotnet/dotnet run

***Configuration**
Build: .NET Core 3.0.0-preview4-27514-07
OS: Linux 4.4.0-17763-Microsoft #253-Microsoft Mon Dec 31 17:49:00 PST 2018
OS Arch: X64

***Working set****
Process.MinWorkingSet: 0
Process.MaxWorkingSet: 9223372036854775807
Process.PeakWorkingSet64: 0
Process.WorkingSet64: 17985536

***Memory*********
Process.PeakPagedMemorySize64: 0
Process.PrivateMemorySize64: 0
Process.NonpagedSystemMemorySize64: 0
Process.PagedMemorySize64: 0
Process.PagedSystemMemorySize64: 0

***Virtual memory
Process.VirtualMemorySize64: 84865474560
Process.PeakVirtualMemorySize64: 0
```

## On Linux running in a Debian container on Windows

```console
C:\testapps\memory>docker run --rm -v c:\testapps\memory:/app -w /app mcr.microsoft.com/dotnet/core-nightly/sdk:3.0 dotnet run

***Configuration**
Build: .NET Core 3.0.0-preview4-27508-07
OS: Linux 4.9.125-linuxkit #1 SMP Fri Sep 7 08:20:28 UTC 2018
OS Arch: X64

***Working set****
Process.MinWorkingSet: 0
Process.MaxWorkingSet: 9223372036854775807
Process.PeakWorkingSet64: 0
Process.WorkingSet64: 25559040

***Memory*********
Process.PeakPagedMemorySize64: 0
Process.PrivateMemorySize64: 0
Process.NonpagedSystemMemorySize64: 0
Process.PagedMemorySize64: 0
Process.PagedSystemMemorySize64: 0

***Virtual memory
Process.VirtualMemorySize64: 2689052672
Process.PeakVirtualMemorySize64: 0

***CGroup*********
CGroup usage: 76718080
CGroup memory limit: 9223372036854771712
```
## On Linux running in a Debian container on Windows -- memory-limited to 100MB

```console
C:\testapps\memory>docker run -m 100mb --rm -v c:\testapps\memory:/app -w /app mcr.microsoft.com/dotnet/core-nightly/sdk:3.0 dotnet run

***Configuration**
Build: .NET Core 3.0.0-preview4-27508-07
OS: Linux 4.9.125-linuxkit #1 SMP Fri Sep 7 08:20:28 UTC 2018
OS Arch: X64

***Working set****
Process.MinWorkingSet: 0
Process.MaxWorkingSet: 9223372036854775807
Process.PeakWorkingSet64: 0
Process.WorkingSet64: 25604096

***Memory*********
Process.PeakPagedMemorySize64: 0
Process.PrivateMemorySize64: 0
Process.NonpagedSystemMemorySize64: 0
Process.PagedMemorySize64: 0
Process.PagedSystemMemorySize64: 0

***Virtual memory
Process.VirtualMemorySize64: 2688528384
Process.PeakVirtualMemorySize64: 0

***CGroup*********
CGroup usage: 72470528
CGroup memory limit: 104857600
```

**Comment by danmosemsft** [comment]
It seems from discussion in various issues and PR's there is not always consensus about how these concepts (originating in Windows) best map to the world of Unix and containers, given there is also a GC that does measurements of its own, and the OS that by some means of its own chooses when to terminate the process. Attempting to twiddle these values in isolation does not seem likely to help (although I agree PeakWorkingSet64 could surely return WorkingSet64 rather than 0...) and/or may make things worse.

I would like us to understand what the customer scenario is and how they expect to achieve it. For example, is the app trying to adjust the size of its own working set (eg., by releasing cached objects) to avoid getting OOM-killer-ed? If so what API are they using today -- or what API would we recommend they look at, and how should they use it?

**Comment by stephentoub** [comment]
> Multiple properties return a value of 0 on Unix.

These are known and have been since the code was written, e.g.
https://github.com/dotnet/corefx/blob/3633ea2c6bf9d52029681efeedd84fd7a06eb6ba/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs#L140-L143
per Dan's explanation. https://github.com/dotnet/corefx/issues/23449 tracks one particular case where there's a concrete suggestion for an API to be used for one of these. If you have suggestions for what APIs should be used to surface the additional information, great, we should fill in as many as we can reasonably.  Otherwise, I'm not sure this is particularly actionable.

> Process.MaxWorkingSet != memory.limit_in_bytes when memory limited

Isn't this tracked by https://github.com/dotnet/corefx/issues/35638?

> Process.WorkingSet64 != memory.usage_in_bytes

We're surfacing the RSS value from procfs:
https://github.com/dotnet/corefx/blob/3633ea2c6bf9d52029681efeedd84fd7a06eb6ba/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs#L137

Can you elaborate on why you believe usage_in_bytes is a better number to surface for WorkingSet64?
https://stackoverflow.com/questions/50865763/memory-usage-discrepancy-cgroup-memory-usage-in-bytes-vs-rss-inside-docker-con

Thanks.

**Comment by luhenry** [comment]
I made some research on the various values reported here, and how they compare to Mono. The following table summaries here the values are read from:


|                              | .NET Core                                         | Mono                            |
| ---------------------------- | ------------------------------------------------- | ------------------------------- |
| `MinWorkingSet` | 0                  | 204800  |
| `MaxWorkingSet` | `/proc/{pid}/stat` | 1413120 |
| `PeakWorkingSet64`           | N/A                                               | `/proc/{pid}/status` → `VmHWM`  |
| `WorkingSet64`               | `/proc/{pid}/stat` * `Environment.SystemPageSize` | `/proc/{pid}/status` → `VmRSS`  |
| `PrivateMemorySize64`        | N/A                                               | `/proc/{pid}/status` → `VmData` |
| `PeakPagedMemorySize64`      | N/A                                               | 0                               |
| `PagedMemorySize64`          | N/A                                               | `/proc/{pid}/status` → `VmSwap` |
| `PagedSystemMemorySize64`    | N/A                                               | `PagedMemorySize64`             |
| `NonpagedSystemMemorySize64` | N/A                                               | 0                               |
| `VirtualMemorySize64`        | `/proc/{pid}/stat`                                | `/proc/{pid}/status` → `VmSize` |
| `PeakVirtualMemorySize64`    | N/A                                               | `/proc/{pid}/status` → `VmPeak` |

I'll update `src/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.cs` to read most of its values from `/proc/{pid}/status` in place of `/proc/{pid}/stat` since it's a superset.

**Comment by abelykh0** [comment]
Would it be possible to fill those values when you can (for the current process or its children), because it is a majority of use cases. For any others feel free to throw `PlatformNotSupportedException`. In any `PlatformNotSupportedException` is better than 0.

```
NonpagedSystemMemorySize64
PagedMemorySize64
PagedSystemMemorySize64
PrivateMemorySize64
```

**Comment by danmosemsft** [comment]
Marking 3.1 to map to porting https://github.com/dotnet/corefx/pull/41122

**Comment by stephentoub** [comment]
Reopening for https://github.com/dotnet/corefx/issues/36086#issuecomment-545081264

**Comment by danmosemsft** [comment]
Ported

**Comment by karelz** [comment]
Fixed in 3.1 in PR #42370 and in 5.0 (master) in PR #41122
